pkg/ovsdb/transact.go

```diff
@@ -0,0 +1,918 @@
+	txn.etcdThen = append(txn.etcdThen, etcdOp)
+	return nil
+}
+
+/* insert */
+func preInsert(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error {
+	if ovsOp.UUIDName == "" {
+		return nil
+	}
+	etcdGetTable(txn, ovsOp.Table)
+	return nil
+}
+
+func doInsert(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error {
+	if ovsOp.UUIDName != "" {
+		for _, row := range txn.etcdCache[ovsOp.Table] {
+			if row["uuid-name"] == ovsOp.UUIDName {
+				err := fmt.Errorf("conflicting uuid-name")
+				ovsResult.Error = err.Error()
+				return err
+			}
+		}
+	}
+	ovsResult.Count = ovsResult.Count + 1
+	uuid := uuid.New() /* generate RFC4122 UUID */
+	return etcdPutRow(txn, ovsOp.Table, uuid.String(), ovsOp.Row)
+}
+
+/* select */
+func preSelect(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error {
+	ok, err := doesWhereContainCondTypeUUID(ovsOp.Where)
+	if err != nil {
+		return err
+	}
+	if ok {
+		return nil
+	}
+	etcdGetTable(txn, ovsOp.Table)
+	return nil
+}
+
+func doSelect(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error {
+	for _, row := range txn.etcdCache[ovsOp.Table] {
+		ok, err := isRowSelectedByWhere(row, ovsOp.Where)
+		if err != nil {
+			return err
+		}
+		if !ok {
+			continue
+		}
+		resultRow := reduceRowByColumns(row, ovsOp.Columns)
+		ovsResult.Count = ovsResult.Count + 1
+		ovsResult.Rows = append(ovsResult.Rows, resultRow)
+
+	}
+	return nil
+}
+
+/* update */
+func preUpdate(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error {
+	return preInsert(txn, ovsOp, ovsResult)
+}
+
+func doUpdate(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error {
+	return doInsert(txn, ovsOp, ovsResult)
+}
+
+/* mutate */
+func preMutate(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error {
+	etcdGetTable(txn, ovsOp.Table)
+	return nil
+}
+
+func doMutate(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error {
+	for uuid, row := range txn.etcdCache[ovsOp.Table] {
+		ok, err := isRowSelectedByWhere(row, ovsOp.Where)
+		if err != nil {
+			return err
+		}
+		if !ok {
+			continue
+		}
+		ovsResult.Count = ovsResult.Count + 1
+		err = RowMutate(&row, ovsOp.Mutations)
+		if err != nil {
+			return err
+		}
+		txn.etcdCache[ovsOp.Table][uuid] = row
+		etcdPutRow(txn, ovsOp.Table, uuid, row)
+	}
+	return nil
+}
+
+/* delete */
+func preDelete(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error {
+	ok, err := doesWhereContainCondTypeUUID(ovsOp.Where)
+	if err != nil {
+		return err
+	}
+	if ok {
+		return nil
+	}
+	etcdGetTable(txn, ovsOp.Table)
+	return nil
+}
+
+func doDelete(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libovsdb.OperationResult) error {
+	for uuid, row := range txn.etcdCache[ovsOp.Table] {
```

Comment:
I will need to warm the cache, so to keep uniformity (I want to avoid multiple paths of execution, will optimize and create different paths only if proven that this causes a performance issue which is tangible)